ocp_utilities/monitoring.py

```diff
@@ -0,0 +1,199 @@
+            namespace=self.namespace, name=self.resource_name, client=self.client
+        )
+
+        secret_name = prometheus_sa.instance.imagePullSecrets[0].name
+        secret = Secret(namespace=self.namespace, name=secret_name, client=self.client)
+
+        token = secret.instance.metadata.annotations["openshift.io/token-secret.value"]
+
+        return {"Authorization": f"Bearer {token}"}
+
+    def _get_response(self, query):
+        response = requests.get(
+            f"{self.api_url}{query}", headers=self.headers, verify=self.verify_ssl
+        )
+
+        try:
+            return json.loads(response.content)
+        except JSONDecodeError as json_exception:
+            LOGGER.error(
+                "Exception converting query response to JSON: "
+                f"exc={json_exception} response_status_code={response.status_code} response={response.content}"
+            )
+            raise
+
+    def query(self, query):
+        """get the prometheus query result
+
+        Args:
+            query(str): promthetheus query string
+
+        Returns:
+            dict: query result
+        """
+        return self._get_response(query=f"{self.api_v1}/query?query={query}")
+
+    def get_alert(self, alert_name):
+        """
+        Get alert by alert anem if it's an active alert
+
+        Args:
+             alert(str): alert name
+
+        Examples:
+             result = prometheus.get_alert(alert='WatchDog')
+
+        Returns:
+             dict: dictiionary containing alert name as key and alert data as value
+        """
+        response = self.alerts
+        LOGGER.info(f" response activer alerts : {response}")
+        matched_alerts = {}
+        alert_list = []
+        for alert in response["data"]["alerts"]:
+            if alert and alert["labels"]["alertname"] == alert_name:
+                alert_list.append(alert)
+        if not alert_list:
+            LOGGER.error(f"No Matching alerts found for alert: {alert_name}")
+        else:
+            matched_alerts[alert_name] = alert_list
+            return matched_alerts
+
+    def alert_sampler(self, alert, timeout):
+        """
+        Sample output for an alert and raises timeout exception
+
+        Args:
+             alert(str): alert name
+             timeout(int): wait time, default is 10 mins
+
+        Raise:
+             TimeoutExpiredError: if alert is not fired before wait_timeout
+        """
+        sampler = TimeoutSampler(
+            wait_timeout=timeout or TIMEOUT_10MIN,
+            sleep=self.scrapeInterval,
+            func=self.get_alert,
+            alert_name=alert,
+        )
+        sample = None
+        try:
+            for sample in sampler:
+                if sample:
+                    for item in sample[alert]:
+                        if item and item["state"] == "firing":
+                            return
+                else:
+                    LOGGER.error("no alert found")
```

Comment:
This is misleading, you may have found a pending alert.

Reply:
i'll update it to "Alert not fired before timeout"

Reply:
That is correct. But I am not talking about case where no sample is found. An alert could be in different states: pending, firing or inactive. https://github.com/prometheus/prometheus/blob/main/rules/alerting.go#L63

if get_alert() returns alerts in pending state, your `if item and item["state"] == "firing":` would still not match. In that case your error message would be misleading. alert_sampler() should not filter firing alerts. Instead we should have a separate functions to get firing_alerts.

Reply:
added another function for firing alerts, updated sampler to make use of get_firing alerts functions